### PR TITLE
Suggestions of new printings for the schemes functionalities

### PR DESCRIPTION
**Pull request by StevellM**
FAO @simonbrandhorst , @HechtiDerLachs , @afkafkafk13 

On demand, I have tried to bring a new (hopefully) consistent printing system to the schemes' functionalities, with respect to the new printing guidelines. Before exposing some examples, here are some comments:
- you will often find a function `_show_semi_compact`.. it is mainly meant for nested printing, to have a detailed printing not mentioning some part of the regular printing (to avoid duplication of information), and allow some off-sets (for the indentation);
- most of the printing functions which can rely on a certain covering allow one to decide on the choice of the covering. If not mention, the methods will take a simplified covering if stored, otherwise the default covering is chosen... as default!
- I have tried to cover all the corner cases, tough there might still be some cases I have missed;
- Not all experimental is yet covered because I do not know much how to handle it, for now. This includes: pullback and pushforward of sheaves, presheaves and hom sheaves (so some sheaves stuff). I would say there are some functionalities missing, and maybe design decisions to be taken before deciding what to do with their printing.
- Some of the printing are still experimental/conditional to changes of printings on the commutative algebra sides (ideals, localized rings, quotient rings, etc...). After these changes, some of choices will make more sense - i.e. printing more details about the patches on which we restrict the object (matter of coordinates names)
- Last but not least: these are suggestions, I am open to discussions, deletions and improvements.

<details><summary>Printing examples 1</summary>
<p>

This example was in the doctests of the previously defined `show_details`, which I have set to the standard printing now (so all blowups are detailed by default). 
```
julia> R, (x,y,z) = QQ["x", "y", "z"];

julia> A3 = Spec(R)
Spectrum
  of multivariate polynomial ring in 3 variables over QQ

julia> I = ideal(R, [x,y,z])
ideal(x, y, z)

julia> bl = blow_up(A3, I) # Detailed
Blow up
  of scheme over QQ covered with 1 affine patch
       spec of multivariate polynomial ring
  in sheaf of ideals with restriction
       patch 1: ideal(x, y, z)
with domain
  scheme over QQ covered with 3 affine patches
    spec of quotient of multivariate polynomial ring
    spec of quotient of multivariate polynomial ring
    spec of quotient of multivariate polynomial ring
and exceptional divisor
  effective cartier divisor defined by
    sheaf of ideals with restrictions
      patch 1: ideal(x)
      patch 2: ideal(y)
      patch 3: ideal(z)

# More compact
julia> (bl,)
(Blow-up: scheme over QQ covered with 3 affine patches -> scheme over QQ covered with 1 affine patch,)

# Supercompact
julia> Base.show(IOContext(stdout, :supercompact => true), bl)
Morphism
```
One can have a taste for some other printings, as `EffectiveCartierDivisor` and `IdealSheaf` by looking at the exceptional divisor and the center of the previous blow up. Important here: for both printing, we need to make sure to show the exact same patches for the ambient scheme and the object defined. Internally, it is made sure that the same coverings is used, and so we have a nice correspondence, allowing one to have the information on which patches they look.

```
julia> E = exceptional_divisor(bl)
Effective cartier divisor
  on scheme over QQ covered with 3 affine patches
       spec of quotient of multivariate polynomial ring
       spec of quotient of multivariate polynomial ring
       spec of quotient of multivariate polynomial ring
defined by
  sheaf of ideals with restrictions
    patch 1: ideal(x)
    patch 2: ideal(y)
    patch 3: ideal(z)

julia> Z = center(bl)
Sheaf of ideals
  on scheme over QQ covered with 1 affine patch
       spec of multivariate polynomial ring
with restriction
  patch 1: ideal(x, y, z)
```
More compactly, on can read

```
julia> (E,)
(Effective cartier divisor on scheme over QQ covered with 3 affine patches,)

julia> Base.show(IOContext(stdout, :supercompact => true), E)
Cartier divisor

# We do not know that Z is prime
julia> (Z,) 
(Sheaf of ideals on scheme over QQ covered with 1 affine patch,)

julia> is_prime(Z)
true

julia> (Z,)
(Sheaf of prime ideals on scheme over QQ covered with 1 affine patch,)

julia> Base.show(IOContext(stdout, :supercompact => true), Z)
Presheaf
```
We can turn `E` into different objects to see other kind of printings

```
julia> D = weil_divisor(E)
Prime weil divisor
  on scheme over QQ covered with 3 affine patches
       spec of quotient of multivariate polynomial ring
       spec of quotient of multivariate polynomial ring
       spec of quotient of multivariate polynomial ring
with coefficients in integer Ring
given as the formal sum of
  1*Sheaf of ideals with restrictions
      patch 1: ideal(x)
      patch 2: ideal(y)
      patch 3: ideal(z)

julia> 2*D
Effective weil divisor
  on scheme over QQ covered with 3 affine patches
       spec of quotient of multivariate polynomial ring
       spec of quotient of multivariate polynomial ring
       spec of quotient of multivariate polynomial ring
with coefficients in integer Ring
given as the formal sum of
  2*Sheaf of ideals with restrictions
      patch 1: ideal(x)
      patch 2: ideal(y)
      patch 3: ideal(z)

julia> -2*D
Weil divisor
  on scheme over QQ covered with 3 affine patches
       spec of quotient of multivariate polynomial ring
       spec of quotient of multivariate polynomial ring
       spec of quotient of multivariate polynomial ring
with coefficients in integer Ring
given as the formal sum of
  -2*Sheaf of ideals with restrictions
      patch 1: ideal(x)
      patch 2: ideal(y)
      patch 3: ideal(z)

```
Note that Weil divisors know whether they are prime or effective, and the printing will adapt so, even in compact printing

```
julia> [D, 2*D, -2*D]
3-element Vector{WeilDivisor{CoveredScheme{QQField}, ZZRing, ZZRingElem}}:
 Prime weil divisor on scheme over QQ covered with 3 affine patches
 Effective Weil divisor on scheme over QQ covered with 3 affine patches
 Weil divisor on scheme over QQ covered with 3 affine patche
```
but the supercompact printing forgets this

```
julia> Base.show(IOContext(stdout, :supercompact => true), D)
Algebraic cycle
```
See also

```
julia> 2*E
Cartier divisor
  on scheme over QQ covered with 3 affine patches
       spec of quotient of multivariate polynomial ring
       spec of quotient of multivariate polynomial ring
       spec of quotient of multivariate polynomial ring
with coefficients in integer Ring
defined by the formal sum of
  2*sheaf of ideals with restrictions
      patch 1: ideal(x)
      patch 2: ideal(y)
      patch 3: ideal(z)

julia> AlgebraicCycle(Z)
Irreducible algebraic cycle 
  on scheme over QQ covered with 1 affine patch
       spec of multivariate polynomial ring
with coefficients in integer Ring
given as the formal sum of
  1*sheaf of prime ideals with restriction
        patch 1: ideal(x, y, z)

julia> 2*AlgebraicCycle(Z)
Effective algebraic cycle 
  on scheme over QQ covered with 1 affine patch
       spec of multivariate polynomial ring
with coefficients in integer Ring
given as the formal sum of
  2*sheaf of prime ideals with restriction
        patch 1: ideal(x, y, z)

julia> -2*AlgebraicCycle(Z)
Algebraic cycle 
  on scheme over QQ covered with 1 affine patch
       spec of multivariate polynomial ring
with coefficients in integer Ring
given as the formal sum of
  -2*sheaf of prime ideals with restriction
         patch 1: ideal(x, y, z)
```

</p>
</details> 

<details><summary>Printing examples 2</summary>
<p>

Here is a bit about the affine/projective varieties, glueings and covered schemes

```
julia> P, x = polynomial_ring(QQ,3);

julia> X = variety(ideal(x))
Affine variety
  in affine 3-space over QQ with coordinates x1, x2, x3
defined by ideal(x1, x2, x3)

julia> (X,)
(V(x1, x2, x3),)

julia> Base.show(IOContext(stdout, :supercompact => true), X)
Scheme

julia> A = ambient_space(X)
Affine space of dimension 3
  over rational field
with coordinates x1, x2, x3

julia> (A,)
(Affine 3-space over QQ with coordinates x1, x2, x3,)

julia> Base.show(IOContext(stdout, :supercompact => true), A)
AA^3

julia> S, y = grade(P);

julia> Y = variety(ideal(y[1:2]))
Projective variety
  in projective 2-space over QQ with coordinates x1, x2, x3
defined by ideal(x2, x1)

julia> (Y,)
(V(x2,x1),)

julia> Base.show(IOContext(stdout, :supercompact => true), Y)
Scheme

julia> B = ambient_space(Y)
Projective space of dimension 2
  over rational field
with homogeneous coordinates x1, x2, x3

julia> (B,)
(Projective 2-space over QQ with coordinates x1, x2, x3,)

julia> Base.show(IOContext(stdout, :supercompact => true), B)
IP^2

julia> C = covered_scheme(Y)
Scheme
  over rational field
with 3 affine patches in its default covering
  spec of quotient of multivariate polynomial ring
  spec of quotient of multivariate polynomial ring
  spec of quotient of multivariate polynomial ring

julia> (C,)
(Scheme over QQ covered with 3 affine patches,)

julia> Base.show(IOContext(stdout, :supercompact => true), C)
Scheme

julia> dc = default_covering(C)
Covering with 3 affine patches
  spec of quotient of multivariate polynomial ring
  spec of quotient of multivariate polynomial ring
  spec of quotient of multivariate polynomial ring

julia> (dc,)
(Covering with 3 affine patches,)

julia> Base.show(IOContext(stdout, :supercompact => true), dc)
Covering

julia> g = collect(values(glueings(dc)))[1]
Glueing
  of  spec of quotient of multivariate polynomial ring
  and spec of quotient of multivariate polynomial ring
along the open subsets
  spec of localized quotient of multivariate polynomial ring
  spec of localized quotient of multivariate polynomial ring
given by
  (x1//x2) -> 1
  (x3//x2) -> (x3//x1)

julia> (g,)
(Glueing of affine patches,)

julia> Base.show(IOContext(stdout, :supercompact => true), g)
Affine glueing
```
and some about morphisms

``` 
# For affine schemes
julia> identity_map(X)
Morphism
  from v(x1, x2, x3)
  to   v(x1, x2, x3)
given by
  x1 -> 0
  x2 -> 0
  x3 -> 0

# For affine spaces
julia> identity_map(A)
Morphism
  from affine 3-space over QQ with coordinates x1, x2, x3
  to   affine 3-space over QQ with coordinates x1, x2, x3
given by
  x1 -> x1
  x2 -> x2
  x3 -> x3

# For projective varieties
# This has no underlying covered scheme morphisms stored, so by default we do not compute anything "new"... once computed, it will be added by itself on the printing
julia> identity_map(Y)
Morphism
  from v(x2,x1)
  to   v(x2,x1)

# For projective spaces
julia> identity_map(B)
Morphism
  from projective 2-space over QQ with coordinates x1, x2, x3
  to   projective 2-space over QQ with coordinates x1, x2, x3

# For coverings
julia> identity_map(dc)
Morphism
  from covering with 3 affine patches
         spec of quotient of multivariate polynomial ring
         spec of quotient of multivariate polynomial ring
         spec of quotient of multivariate polynomial ring
  to   covering with 3 affine patches
         spec of quotient of multivariate polynomial ring
         spec of quotient of multivariate polynomial ring
         spec of quotient of multivariate polynomial ring
given by
  patch 1:
    (x2//x1) -> 0
    (x3//x1) -> 0
  ----------------------------------------
  patch 2:
    (x1//x2) -> 0
    (x3//x2) -> 0
  ----------------------------------------
  patch 3:
    (x1//x3) -> 0
    (x2//x3) -> 0

# For covered schemes
julia> identity_map(C)
Morphism
  from scheme over QQ covered with 3 affine patches
         spec of quotient of multivariate polynomial ring
         spec of quotient of multivariate polynomial ring
         spec of quotient of multivariate polynomial ring
  to   scheme over QQ covered with 3 affine patches
         spec of quotient of multivariate polynomial ring
         spec of quotient of multivariate polynomial ring
         spec of quotient of multivariate polynomial ring
given by
  patch 1:
    (x2//x1) -> 0
    (x3//x1) -> 0
  ----------------------------------------
  patch 2:
    (x1//x2) -> 0
    (x3//x2) -> 0
  ----------------------------------------
  patch 3:
    (x1//x3) -> 0
    (x2//x3) -> 0

```
Note that on supercompact printings, all the previous show just
```
julia> Base.show(IOContext(stdout, :supercompact => true), identity_map(X))
Morphism
```

</p>
</details>

### CODE REVIEW

**Comment by StevellM** [comment]
I will run a fix for the doctests once we will have agreed on what to keep and what to change among the new printing methods.

**Comment by simonbrandhorst** [comment]
Looks MUCH better! Thank you. I noticed that some v needs to be a V e.g. in
```
Morphism
  from v(x1, x2, x3)
  to   v(x1, x2, x3)
```

**Comment by afkafkafk13** [comment]
The overall impression is much better than before. Thanks a lot.

I have a small comment on one details:
The printing of the last three lines in the output below (which is just one example of printing whenever coverings are involved) is still not very useful for the user.  Depending on the circumstances it is either too much or not enough. 

- too much in the sense that it does not contain any non-trivial information (our schemes are designed in the multivariate polynomial ring world anyway) 
- not enough to eyeball the charts as sophisticated users do, when they are looking for a suitable chart to focus on

```
Scheme
  over rational field
with 3 affine patches in its default covering
  spec of quotient of multivariate polynomial ring
  spec of quotient of multivariate polynomial ring
  spec of quotient of multivariate polynomial ring
```

**Comment by StevellM** [comment]
Thanks @afkafkafk13 for your comments! In fact, I was not sure to keep track of the charts in the printing of a scheme alone. Though, I felt it was important whenever another object is defined on the scheme.

I just want to point out something: we would need to revise the printings of the polynomial rings, so ultimately the chart won't show "... quotient of multivariate polynomial ring", but something more explicit i.e. with coordinates. Though I could not make it for now since there are more things to be discussed about it. The final idea would be to give details about the coordinates of the different charts that could be used for local descriptions of maps, presheaves, glueings, etc...

In any case, I would agree to remove the printing of the charts for a "bold" scheme as in the example you have shown, for now, but maybe it could be worth to have such details whenever we have new printings for the (quotient/localized) polynomial rings.

Or do you have any other suggestion ? What would you expect to see as non-trivial in the printing of the charts ?

**Comment by afkafkafk13** [comment]
I fully agree: 
For now, this solution for the schemes is as good as any, but we need an improved way of printing for quotients of polynomial rings allowing to quickly read off essential information.

**Comment by simonbrandhorst** [comment]
As for the printing of the charts. Printing the coordinate ring will always be suboptimal since such a ring prints its base ring and that will be the same base_ring as that of the scheme. Therefore we will have lots of redundant information. 
It might be better to do some custom printing here. 
I think the following information is relevant:
- coordinates
- the `ambient_closure_ideal` i.e. the relations 
- if it is an open subset, print the complement taken

**Comment by afkafkafk13** [comment]
Just another observation: It might be worth printing the center of a blow-up, as a user might have used several blow-up maps and the easiest 'key' to find the right one is the center.

**Comment by StevellM** [comment]
@simonbrandhorst  Good, so this PR will take a bit more time then, I will try to apply your suggestions and do some custom printings for the charts! Could you develop what you mean by "if it is an open subset, print the complement taken" ?

**Comment by afkafkafk13** [comment]
The meaning of 'if it is an open' is the following: 
There might be charts which are of the structure D(f) \subset A^n for some polynomial f. In this case, it is important for the user to know that the chart is not the whole A^n, but some D(f), and hence to specify the polynomial f. 
You might also find settings where we do not store f as a single polynomial, but a factorization thereof -- in this case, the factors should be returned.

**Comment by StevellM** [comment]
@afkafkafk13 @HechtiDerLachs after discussion with Simon, we keep for now the printing almost as is: we do not yet print the details of the charts but at least the coordinates in which they are defined. We do so to see more details on the coordinates, and it would have to be decided later what to do as custom printings for the description of the charts (maybe in terms of `V(-)` and `D(-)` )
My main comment has been adapted accordingly to the new changes, so please have a look again to the same examples (I have added two more for zero Weil divisors/Algebraic cycles)

@HereAround To avoid conflict with the doctests from the experimental in `FTheoryTools`, I had to add two "fake" printing methods. In fact, the `ToricCoveredScheme` in FTheoryTools are a kind of `CoveredScheme` and since the package has not completely yet adapted the new printing conventions, I had some conflicts issues. (I have also found a bug somewhere, I will reach you about this)

**Comment by afkafkafk13** [comment]
Printing looks good to me. 
Nevertheless, we should keep in mind that we want a mathematically more informative printing V(...) , D(...) or V(...) \cap D(...) printing for the charts further down the road.

**Comment by simonbrandhorst** [comment]
> Printing looks good to me. Nevertheless, we should keep in mind that we want a mathematically more informative printing V(...) , D(...) or V(...) \cap D(...) printing for the charts further down the road.

Agreed. But what do we do if V(... ) is not nice but rather big (as usual). Embrace it and live with the fact (like we do for ideals?)

File: src/AlgebraicGeometry/Schemes/AffineVariety/Objects/Constructors.jl [hunk 1]
```diff
@@ -32,7 +32,7 @@ julia> R, (x,y) = QQ[:x,:y]
 
 julia> variety(ideal([x,y]))
 Affine variety
-  in affine 2-space over QQ with coordinates x, y
+  in 𝔸² over QQ with coordinates x, y
```
Comment (simonbrandhorst) [inline src/AlgebraicGeometry/Schemes/AffineVariety/Objects/Constructors.jl]: This looks like some `is_unicode_enabled` is not checked?

File: src/AlgebraicGeometry/Schemes/ProjectiveAlgebraicSet/Objects/Constructors.jl [hunk 2]
```diff
@@ -97,7 +97,7 @@ julia> (s0,s1) = homogeneous_coordinates(P1);
 
 julia> X = algebraic_set((s0^2+s1^2)*s1)
 Algebraic set
-  in projective 1-space over QQ with coordinates s0, s1
+  in ℙ¹ over QQ
```
Comment (simonbrandhorst) [inline src/AlgebraicGeometry/Schemes/ProjectiveAlgebraicSet/Objects/Constructors.jl]: same here

File: src/AlgebraicGeometry/Schemes/AffineSchemes/Objects/Methods.jl [hunk 3]
```diff
@@ -25,21 +25,31 @@ end
...
 
-
 ########################################################
 # (2) Display
 ########################################################
 
+function Base.show(io::IO, ::MIME"text/plain", X::AbsSpec)
+  io = pretty(io)
+  println(io, "Spectrum")
+  print(io, Indent(), "of ", Lowercase(), OO(X))
```
Comment (simonbrandhorst) [inline src/AlgebraicGeometry/Schemes/AffineSchemes/Objects/Methods.jl]: Let us go for the detailed printing of `OO(X)` here.

File: src/AlgebraicGeometry/Schemes/ProjectiveAlgebraicSet/Objects/Constructors.jl (hunk 2 above)
Comment (simonbrandhorst) [reply src/AlgebraicGeometry/Schemes/ProjectiveAlgebraicSet/Objects/Constructors.jl]: or perhaps the patchbot has it set to true?

**Comment by simonbrandhorst** [comment]
Could you please add some tests for the printing? You could e.g. add doctests for the show methods. 
The last time I programmed the printing, I messed up a lot of methods and only writing the test, made it all work. 
So far we have little doctests because the printing was so horrible.

File: src/AlgebraicGeometry/Schemes/AffineVariety/Objects/Constructors.jl (hunk 1 above)
Comment (StevellM) [reply src/AlgebraicGeometry/Schemes/AffineVariety/Objects/Constructors.jl]: Yes it is, I did not change the condition on the unicode checked

File: src/AlgebraicGeometry/Schemes/ProjectiveAlgebraicSet/Objects/Constructors.jl (hunk 2 above)
Comment (StevellM) [reply src/AlgebraicGeometry/Schemes/ProjectiveAlgebraicSet/Objects/Constructors.jl]: It looks like my machine allows it and so everytime it makes it appear. I will have to deal again with those doctests fixes..

**Comment by afkafkafk13** [comment]
> > Printing looks good to me. Nevertheless, we should keep in mind that we want a mathematically more informative printing V(...) , D(...) or V(...) \cap D(...) printing for the charts further down the road.
> 
> Agreed. But what do we do if V(... ) is not nice but rather big (as usual). Embrace it and live with the fact (like we do for ideals?)

I'd say live with it. 

An alternative could be the following (but that would mean quite some work without too much benefit):
Do we have some notions of how many terms this element has? If it is beyond some threshold, we could just print the leading term of it, a '+' and '...'. 
Example: x^2+y^3+z^5 would print as z^5+... for degrevlex (i.e. default_ordering).

**Comment by HechtiDerLachs** [comment]
This all looks good to me. Thank you very much for all the work you put into this!

**Comment by afkafkafk13** [comment]
This looks fine to me, but it still needs doctests to make sure that everyone touching any printing (e.g. in the rings) becomes aware of the effect in the schemes...
Say when you have the tests ready and you will get an approval.

**Comment by StevellM** [comment]
Good, then if we forget about the nightly tests, which are currently buggy, everything should look good. I just realised that my initial comment has never been updated. @afkafkafk13 would you still like to see some examples ? I could re-try to update the comment, or I can point out at some doctests I have added.

**Comment by afkafkafk13** [comment]
> Good, then if we forget about the nightly tests, which are currently buggy, everything should look good. I just realised that my initial comment has never been updated. @afkafkafk13 would you still like to see some examples ? I could re-try to update the comment, or I can point out at some doctests I have added.

I already looked at the doctests you added with the last PR and they seem to cover a lot of ground (by the nature of the objects they cover). So at this point I am fine, as soon as the tests succeed (of course ignoring the broken nightly ones)

**Review by afkafkafk13 — APPROVED** [review]

File: src/AlgebraicGeometry/Schemes/AffineAlgebraicSet/Objects/Types.jl [hunk 4]
```diff
@@ -1,11 +1,49 @@
...
+# Examples
+
+```jldoctest
+julia> A = affine_space(QQ, [:x,:y]);
+
+julia> (x, y) = coordinates(A);
+
+julia> X = algebraic_set(ideal([y - x^2]))
+Affine algebraic set
+  in 𝔸² over QQ with coordinates x, y
```
Comment (simonbrandhorst) [inline src/AlgebraicGeometry/Schemes/AffineAlgebraicSet/Objects/Types.jl:16]: is unicode allowed?

File: src/AlgebraicGeometry/Schemes/AffineSchemes/Morphisms/Attributes.jl [hunk 5]
```diff
@@ -213,11 +239,22 @@ julia> (x1,x2,x3) = gens(R)
...
+Morphism
+  from [x1, x2, x3]  spec of quotient of multivariate polynomial ring
+  to   [x1, x2, x3]  affine 3-space over QQ
+given by the pullback function
+  x1 -> 0
+  x2 -> x2
+  x3 -> x3
+
+julia> graph(f)
+(Spec of quotient of multivariate polynomial ring, Morphism: spec of quotient of multivariate polynomial ring -> spec of quotient of multivariate polynomial ring, Morphism: spec of quotient of multivariate polynomial ring -> affine 3-space over QQ with coordinates x1, x2, x3)
```
Comment (simonbrandhorst) [inline src/AlgebraicGeometry/Schemes/AffineSchemes/Morphisms/Attributes.jl:257]: In one line mode any commas should be enclosed by ( ) This is to make sure that we can read a tuple. 
E.g. here: are x1, x2, x3) components of the tuple that is returned?

File: src/AlgebraicGeometry/Schemes/CoveredSchemes/Objects/Attributes.jl [hunk 6]
```diff
@@ -156,11 +201,67 @@ end
   return domain(inc), inc
 end
 
+@doc raw"""
+    singular_locus(X::AbsCoveredScheme) -> AbsCoveredScheme, AbsCoveredSchemeMorphism
+
+Return the singular locus of `X` as a covered scheme.
+
+For the singular locus of the reduced scheme induced by `X`, please use
+`singular_locus_reduced`.
```
Comment (simonbrandhorst) [inline src/AlgebraicGeometry/Schemes/CoveredSchemes/Objects/Attributes.jl:210]: I seem to recall that `singular_locus_reduced` returns the singular locus as a reduced scheme?
(and not the singular locus of the reduced scheme). Please check and adjust accordingly.

File: src/AlgebraicGeometry/Schemes/SpecOpen/Objects/Constructors.jl [hunk 7]
```diff
@@ -62,12 +103,86 @@ end
...
 # Additional constructors                                              #
 ########################################################################
+
+@doc raw"""
+    product(U::SpecOpen, Z::AbsSpec) -> SpecOpen, SpecOpenMor, SpecOpenMor
+
+Given a Zariski open subset `U` of an affine scheme `X`, complement to
+a subscheme `Y` of `X`, and given an affine scheme `Z`, return the product
+scheme $V := U \times Z$ as a Zariski open subset of the affine scheme
+$X\times Z$ complement to the closed subschme $Y\times Z$.
```
Comment (simonbrandhorst) [inline src/AlgebraicGeometry/Schemes/SpecOpen/Objects/Constructors.jl:119]:
```suggestion
$X\times Z$ complement to the closed subscheme $Y\times Z$.
```

**Comment by simonbrandhorst** [comment]
Thanks for this nice piece of work!
What I am a little unsure, how does it work with the offsets & indentations etc. if we print some more ugly real world examples?

File: src/AlgebraicGeometry/Schemes/AffineAlgebraicSet/Objects/Types.jl (hunk 4 above)
Comment (StevellM) [reply src/AlgebraicGeometry/Schemes/AffineAlgebraicSet/Objects/Types.jl:16]: I have no clue... It fixed it by itself, and the tests are no complaining so it seems to. I do not know why it does so, I disallowed it locally

File: src/AlgebraicGeometry/Schemes/AffineSchemes/Morphisms/Attributes.jl (hunk 5 above)
Comment (StevellM) [reply src/AlgebraicGeometry/Schemes/AffineSchemes/Morphisms/Attributes.jl:257]: Ah good point, I should removed the coordinates here

File: src/AlgebraicGeometry/Schemes/AffineAlgebraicSet/Objects/Types.jl (hunk 4 above)
Comment (simonbrandhorst) [reply src/AlgebraicGeometry/Schemes/AffineAlgebraicSet/Objects/Types.jl:16]: If the patchbots are happy and it behaves as intended this is o.k.

**Comment by simonbrandhorst** [comment]
We keep the nitpicks for another pr.

**Pull request merged**